### PR TITLE
fixing the issue #629

### PR DESCRIPTION
**Pull request by faezeh-R**

### CODE REVIEW

File: src/pkcs11/slot.c [hunk 1]
```diff
@@ -454,9 +455,11 @@ CK_RV slot_token_removed(CK_SLOT_ID id)
 CK_RV slot_find_changed(CK_SLOT_ID_PTR idp, int mask)
 {
 	unsigned int i;
+	CK_RV rv;
 	LOG_FUNC_CALLED(context);
 
-	card_detect_all();
+	rv = card_detect_all();
+	if(rv == CKR_TOKEN_NOT_PRESENT ) LOG_FUNC_RETURN(context, CKR_TOKEN_NOT_PRESENT);
```
Comment (frankmorgner) [inline src/pkcs11/slot.c]: this doesn't work. card_detect_all will always return `CKR_OK`!

File: src/pkcs11/pkcs11-global.c [hunk 2]
```diff
@@ -656,9 +656,13 @@ CK_RV C_WaitForSlotEvent(CK_FLAGS flags,   /* blocking/nonblocking flag */
 		mask |= SC_EVENT_READER_EVENTS;
 	}
 
+
+	sc_ctx_detect_readers(context);
+
 	rv = slot_find_changed(&slot_id, mask);
-	if ((rv == CKR_OK) || (flags & CKF_DONT_BLOCK))
+	if (rv == CKR_OK   ||  (rv != CKR_TOKEN_NOT_PRESENT  && (flags & CKF_DONT_BLOCK) ))
```
Comment (frankmorgner) [inline src/pkcs11/pkcs11-global.c]: this doesn't work. card_detect_all will never return `CKR_TOKEN_NOT_PRESENT`!

Comment (faezeh-R) [inline src/pkcs11/pkcs11-global.c]: I changed the return of carddetectAll in slot.c

**Comment by frankmorgner** [comment]
I think this is the wrong fix. This may work with firefox, but I it is not compliant with PKCS#11. The return code is not defined in this context, it does not set the slot that changed.

**Comment by dengert** [comment]
I agree with @frankmorgner 

 An OpenSC SPY trace would be most helpful to see what FireFox is actually doing.

**Comment by faezeh-R** [comment]
But this solution works in firefox and pkcs11-tools.exe , both of them not just in firefox.
I traced the code and saw that when the reader doesn't found, card_detect function returns the error of "READER_NOT_FOUND" but the function of card_detect_all always returns CKR_OK.
then In function of slot_find_changed for doesn't detecting the event of reader, It returns CKR_NO_EVENT.
finally in function of C_WaitForSlotEvent, the codes with label again never execute.

**Comment by frankmorgner** [comment]
I was not able to reproduce this behavior with pkcs11-tool, only with FF.

If the reader was removed its slot *should* be marked with a removed token and a changed state. However, currently `slot_find_changed` does *not recognize* the changes state because the data structure "slot" has already been removed from the internal memory in `card_detect_all`.

The real fix would be to keep the slot alive though the reader has been removed and to propagate the changed state (token removed). You could use https://github.com/frankmorgner/OpenSC/commit/87327030b42641815a6bdc0f8491a1e7b6f32b20 as starting point, but beware it is neither functional (events are still not propagated correctly) nor correct (segfaults in FF)...

**Comment by frankmorgner** [comment]
as pointed out, above a fix should take a different path

**Pull request closed without merging**